web/modules/custom/joinup_core/src/Plugin/Field/FieldWidget/StateMachineButtons.php

```diff
@@ -0,0 +1,192 @@
+<?php
+
+namespace Drupal\joinup_core\Plugin\Field\FieldWidget;
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Field\FieldItemListInterface;
+use Drupal\Core\Field\Plugin\Field\FieldWidget\OptionsSelectWidget;
+use Drupal\Core\Form\FormStateInterface;
+
+/**
+ * Plugin implementation of the 'state_machine_buttons' widget.
+ *
+ * @FieldWidget(
+ *   id = "state_machine_buttons",
+ *   label = @Translation("State machine buttons"),
+ *   field_types = {
+ *     "state"
+ *   },
+ *   multiple_values = TRUE
+ * )
+ */
+class StateMachineButtons extends OptionsSelectWidget {
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function defaultSettings() {
+    return [
+      'use_transition_label' => FALSE,
+    ] + parent::defaultSettings();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function settingsForm(array $form, FormStateInterface $form_state) {
+    $element = [];
+
+    $element['use_transition_label'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Use transition labels'),
+      '#description' => $this->t('Leave unchecked to use <em>Save as</em> followed by the state label.'),
+      '#default_value' => $this->getSetting('use_transition_label'),
+    ];
+
+    return $element;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function settingsSummary() {
+    $summary = [];
+
+    $use_transition_label = $this->getSetting('use_transition_label');
+    if ($use_transition_label) {
+      $summary[] = t('Use transition labels');
+    }
+
+    return $summary;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
+    $element = parent::formElement($items, $delta, $element, $form, $form_state);
+
+    if ($this->getSetting('use_transition_label')) {
+      $element['#options'] = $this->replaceStateLabelsWithTransitionLabels($element['#options'], $items);
+    }
+
+    // Pass the label settings to the process callback.
+    $element['#use_transition_label'] = $this->getSetting('use_transition_label');
+
+    // Leave the field in place for validation purposes.
+    $element['#access'] = FALSE;
+
+    // Add a process callback to add the buttons in the form actions.
+    // @see \Drupal\content_moderation\Plugin\Field\FieldWidget\ModerationStateWidget::formElement()
+    $element['#process'][] = [get_called_class(), 'processActions'];
```

Comment:
Haha this doesn't exist yet in our code base, but it will be when we upgrade to Drupal 8.2 :)